flang/lib/Optimizer/Transforms/RewriteLoop.cpp

```diff
@@ -25,34 +25,119 @@ using namespace fir;
 
   mlir::LogicalResult
   matchAndRewrite(LoopOp loop, mlir::PatternRewriter &rewriter) const override {
-    auto loc = loop.getLoc();
-    auto low = loop.lowerBound();
-    auto high = loop.upperBound();
-    auto step = loop.step();
-    assert(low && high && step);
-    // ForOp has different bounds semantics. Adjust upper bound.
-    auto adjustUp = rewriter.create<mlir::AddIOp>(loc, high, step);
-    auto f = rewriter.create<mlir::scf::ForOp>(loc, low, adjustUp, step);
-    f.region().getBlocks().clear();
-    rewriter.inlineRegionBefore(loop.region(), f.region(), f.region().end());
-    rewriter.eraseOp(loop);
+    Location loc = loop.getLoc();
+
+    // Create the start and end blocks that will wrap the LoopOp with an
+    // initalizer and an end point
+    mlir::Block *initBlock = rewriter.getInsertionBlock();
+    auto initPos = rewriter.getInsertionPoint();
+    mlir::Block *endBlock = rewriter.splitBlock(initBlock, initPos);
+
+    // Split the first LoopOp block in two parts. The part before will be the
+    // conditional block since it already has the induction variable and
+    // loop-carried values as arguments.
+    mlir::Block *conditionalBlock = &loop.region().front();
+    mlir::Block *firstBlock = rewriter.splitBlock(conditionalBlock,
+                                                  conditionalBlock->begin());
+    mlir::Block *lastBlock = &loop.region().back();
+
+    // Move the blocks from the LoopOp between initBlock and endBlock
+    rewriter.inlineRegionBefore(loop.region(), endBlock);
+
+    // Get loop values from the LoopOp
+    mlir::Value low = loop.lowerBound();
+    mlir::Value high = loop.upperBound();
+    mlir::Value step = loop.step();
+    if (!low || !high) {
+      return failure();
+    }
+
+
+    // Initalization block
+    rewriter.setInsertionPointToEnd(initBlock);
+    mlir::Value zero = rewriter.create<mlir::ConstantIndexOp>(loc, 0);
+    mlir::Value one = rewriter.create<mlir::ConstantIndexOp>(loc, 1);
+    mlir::Value diff = rewriter.create<mlir::SubIOp>(loc, high, low);
+    mlir::Value distance = rewriter.create<mlir::AddIOp>(loc, diff, step);
+    mlir::Value normalizedDistance =
+      rewriter.create<mlir::SignedDivIOp>(loc, distance, step);
+    mlir::Type storeType = normalizedDistance.getType();
+    mlir::Value tripCounter =
+      rewriter.create<fir::AllocaOp>(loc, storeType, llvm::None);
+    rewriter.create<fir::StoreOp>(loc, normalizedDistance, tripCounter);
+
+    SmallVector<mlir::Value, 8> loopOperands;
+    loopOperands.push_back(low);
+    mlir::ValueRange operands = loop.getIterOperands();
+    loopOperands.append(operands.begin(), operands.end());
+
+    // TODO: replace with a command line flag
+    // onetrip flag determines whether loop should be executed once, before
+    // conditionals are checked
+    static const bool onetrip = false;
+    if (onetrip) {
+      rewriter.create<BranchOp>(loc, firstBlock, ArrayRef<Value>());
```

Comment:
How are the loop operands propagated to the first-block in the onetrip case ?
I am not an mlir CFG expert at all, my understanding is that in the none-onetrip case the `conditionalBlock` dominates `firstBlock` so its fine for `firstBlock` to use `conditionalBlock` arguments, but in the onetrip case, that seems to not be true anymore, hence my question.
I saw that enabling the one-trip is a TODO, so that is probably just fine for now and maybe there is no issue here anyway.

Reply:
I think the one trip case should retain the loop structure as is, but add an mlir::SelectOp to force the loop count to be at least 1.
Note that in merging this, I also took the liberty of promoting the trip count to a proper SSA-value.

Reply:
I was unsure how to maintain the loop operands in an elegant way between both the onetrip and non-onetrip cases so I simply threw that branch in as a reminder that the onetrip option should be added in the future. As is it will not even compile if onetrip is true.